### PR TITLE
[#8] [Backend] As a user, I can login

### PR DESCRIPTION
**Pull request by ducbm051291**
#8 

## What happened

The user should be able to login by calling Login API.
 
## Insight

- Create GranType Constant
- Create Token, APIToken
- Create KeyChainKey and KeyChain
- Create UseDefaultsKey and UserDefaultsKeyManager
- Create LoginUseCase
- Create AuthenticationRepository
- Create Resolver+Data extension to register services, use case for injection.
- Create SessionRepository
- Create StoreTokenUseCase
- Create Resolver+Domain extension and Resolver+ResolverRegistering to register injection services, domains, and data
- Create extension for publisher to convert to Driver and Observable
- Update typealias Driver
- Create KeychainProtocolMock
- Create APIToken+Dummy
- Create Resolver extension in Test to register injections
- Write UnitTest for AuthenticationRepository
- Write UnitTest for SessionRepository
- Write UnitTest for LoginUseCase
- Write UnitTest for StoreTokenUseCase
 
## Proof Of Work

Screenshot 2022-11-18 at 14 06 27

https://user-images.githubusercontent.com/24598204/202644406-dd59b0e0-a93e-4244-b1af-3eb51fbef3a7.mp4

### CODE REVIEW

File: Surveys/Sources/Data/Storages/Keychain/Keychain.swift [hunk 1]
```diff
@@ -0,0 +1,69 @@
+//
+//  KeyChain.swift
```
Comment (blyscuit) [inline Surveys/Sources/Data/Storages/Keychain/Keychain.swift:2]: This does not match the file name 🤔

File: SurveysTests/Sources/Injection/Resolver+Domain.swift [hunk 2]
```diff
@@ -0,0 +1,9 @@
+//
+//  Resolver+Domain.swift
+//  SurveysTests
+//
+//  Created by David Bui on 17/11/2022.
+//  Copyright © 2022 Nimble. All rights reserved.
+//
+
+import Foundation
```
Comment (blyscuit) [inline SurveysTests/Sources/Injection/Resolver+Domain.swift:9]: This file is empty?

File: SurveysTests/Sources/Specs/Data/Repositories/AuthenticationRepositorySpec.swift [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                            expectation.fulfill()
+                        }
+                        .store(in: &cancelBag)
+
+                    wait(for: [expectation], timeout: 1)
+                }
+
+                context("when networkAPI.performRequest emits failure") {
+
+                    let expectation = XCTestExpectation(description: "Emit error")
```
Comment (blyscuit) [inline SurveysTests/Sources/Specs/Data/Repositories/AuthenticationRepositorySpec.swift:79]: You can wrap the output of sink with the extension `awaitPublisher` from https://www.swiftbysundell.com/articles/unit-testing-combine-based-swift-code/. Then you can use it in Quick/Nimble. That way you will have cleaner test code.

**Review by blyscuit — COMMENTED** [review]
Functionally, looks good.
This is a large PR that can be separated to smaller parts, out of the box I can see:
1. Set up local storage.
2. Login network  backend
3. Login storage backend

**Comment by ducbm051291** [comment]
> Functionally, looks good. This is a large PR that can be separated to smaller parts, out of the box I can see:
> 
> 1. Set up local storage.
> 2. Login network  backend
> 3. Login storage backend

Sound goods 👍

**Pull request closed without merging**